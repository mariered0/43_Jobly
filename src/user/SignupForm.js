import React, { useContext, useState } from "react";
import { useHistory } from "react-router-dom";
import userContext from "./UserContext";
import { Form, FormGroup, Label, Input, Card, CardBody, Button, Alert, List} from "reactstrap";

const SignupForm = () => {
    const { token, signup } = useContext(userContext)
    const [errors, setErrors] = useState(null);
    const history = useHistory();

    const INITIAL_STATE = {
        username: "",
        password: "",
        firstName: "",
        lastName: "",
        email: ""
    }

    const [formData, setFormData] = useState(INITIAL_STATE);

    const handleChange = (e) => {
        const {name, value} = e.target;
        setFormData((formData) => ({
            ...formData,
            [name]: value
        }));
    }
    
    async function handleSubmit (e) {
        e.preventDefault();
        const msg = await signup(formData);
        if (msg.success) {
          history.push('/');
        } else{
          setErrors(errors => [msg.e]);
        }
        setFormData(INITIAL_STATE);
    }

  return (
    <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
      <h2 className="mb-3">Sign up</h2>
      <Card>
        <CardBody>
          <Form>
            <FormGroup>
              <Label for="username" className="mt-3">Username</Label>
              <Input 
                id="username"
                name="username"
                type="text"
                
                onChange={handleChange} />
              
              <Label for="password" className="mt-3">Password</Label>
              <Input 
                id="password"
                name="password"
                type="password"
                autoComplete="on"
                onChange={handleChange} />
              
              <Label for="firstName" className="mt-3">First name</Label>
              <Input 
                id="firstName"
                name="firstName"
                type="text"
                onChange={handleChange} />
              
              <Label for="lastName" className="mt-3">Last name</Label>
              <Input 
                id="lastName"
                name="lastName"
                type="text"
                onChange={handleChange} />
              
              <Label for="email" className="mt-3">Email</Label>
              <Input 
                id="email"
                name="email"
                type="email"
                onChange={handleChange} />
        
            </FormGroup>
            
            { errors &&
            <Alert color="danger">
              {errors.map((error, idx) => (
                <p key={idx}>{error}</p>
              ))}
            </Alert>
            }

            <Button onClick={handleSubmit}>Submit</Button>
          </Form>
        </CardBody>
      </Card>
    </div>
  );
};

export default SignupForm;
