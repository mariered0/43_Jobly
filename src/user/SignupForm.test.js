import React from "react";
import { MemoryRouter } from "react-router";
import { render, fireEvent } from "@testing-library/react";
import SignupForm from "./SignupForm";
import JoblyApi from "../api/api";

import UserContext from "../user/UserContext";

const token = JoblyApi.token;
const signup = jest.fn();

it("renders without crashing", () => {
  render(
    <MemoryRouter>
      <UserContext.Provider value={{ token, signup }}>
        <SignupForm />
      </UserContext.Provider>
    </MemoryRouter>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={{ token, signup }}>
        <SignupForm />
      </UserContext.Provider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("has the login form and accept input", () => {
  const { getByLabelText } = render(
    <MemoryRouter>
      <UserContext.Provider value={{ token, signup }}>
        <SignupForm />
      </UserContext.Provider>
    </MemoryRouter>
  );
  const username = getByLabelText('Username');
  const password = getByLabelText('Password');
  const firstName = getByLabelText('First name');
  fireEvent.change(username, {target: {value: 'abc'}});
  fireEvent.change(password, {target: {value: '123456'}});
  fireEvent.change(firstName, {target: {value: 'marie'}});

  expect(username.value).toBe('abc');
  expect(password.value).toBe('123456');
  expect(firstName.value).toBe('marie');
});