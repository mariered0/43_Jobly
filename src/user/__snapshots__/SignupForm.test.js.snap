// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
  >
    <h2
      class="mb-3"
    >
      Sign up
    </h2>
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="mt-3 form-label"
              for="username"
            >
              Username
            </label>
            <input
              class="form-control"
              id="username"
              name="username"
              type="text"
            />
            <label
              class="mt-3 form-label"
              for="password"
            >
              Password
            </label>
            <input
              autocomplete="on"
              class="form-control"
              id="password"
              name="password"
              type="password"
            />
            <label
              class="mt-3 form-label"
              for="firstName"
            >
              First name
            </label>
            <input
              class="form-control"
              id="firstName"
              name="firstName"
              type="text"
            />
            <label
              class="mt-3 form-label"
              for="lastName"
            >
              Last name
            </label>
            <input
              class="form-control"
              id="lastName"
              name="lastName"
              type="text"
            />
            <label
              class="mt-3 form-label"
              for="email"
            >
              Email
            </label>
            <input
              class="form-control"
              id="email"
              name="email"
              type="email"
            />
          </div>
          <button
            class="btn btn-secondary"
            type="button"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
